source/docs/software/commandbased/organizing-command-based.rst

```diff
@@ -0,0 +1,259 @@
+
+        public RunIntakeCommand(Intake intake) {
+            this.m_intake = intake;
+            addRequirements(intake);
+        }
+
+        @Override
+        public void initialize() {
+            m_intake.setPercent(1);
+        }
+
+        @Override
+        public void end(boolean interrupted) {
+            m_intake.setPercent(0);
+        }
+
+        @Override
+        public boolean isFinished() {
+            return false;
+        }
+    }
+
+  .. code-tab:: c++
+
+    // TODO
+
+This, however, is just as cumbersome. The only two lines that really matter in this entire file are the two calls to ``intake.setPercent``, yet there are over 20 lines of boilerplate code! Not to mention, doing this for a lot of robot actions quickly clutters up a robot project with dozens of small files. Nevertheless, this might feel more "natural," particularly for programmers who don't feel confident with Java and want to stick closely to an object-oriented model.
+
+Factory Methods
+^^^^^^^^^^^^^^^
+
+A useful middle ground between these two extremes is using a `factory method <https://www.tutorialspoint.com/design_pattern/factory_pattern.htm>`__. A factory method is a method that, each time it is called, returns a new object according to some specification.
+
+A command like the intake-running command is conceptually related to exactly one subsystem: the ``Intake``. As such, it makes sense to put a ``commandRun`` method as an instance method of the ``Intake`` class:
+
+.. tabs::
+
+  .. code-tab:: java
+
+    public class Intake {
+        // [code for motor controllers, configuration, etc.]
+        // ...
+
+        public Command commandRun() {
+            return new StartEndCommand(() -> this.set(1), () -> this.set(0), this);
+        }
+    }
+
+  .. code-tab:: c++
+
+    // TODO
+
+Notice how since we are in the ``Intake`` class, we no longer refer to ``intake``; instead, we use the ``this`` keyword to refer to the current instance.
+
+.. warning:: To preserve encapsulation, avoid referring to ``private`` variables and methods of the ``Intake`` class from within the ``commandRun`` method, even though they are technically accessible. (For instance, the ``commandRun`` method should use the subsystem's public ``set`` method instead of interfacing directly with the motor controllers.)
+
+Using this new factory method in command groups and button bindings is highly expressive:
+
+
+.. tabs::
+
+  .. code-tab:: java
+
+    intakeButton.whileHeld(intake.commandRun());
+
+    Command intakeAndShoot = intake.commandRun().alongWith(new RunShooter());
+
+    Command autonomousCommand = new SequentialCommandGroup(
+        intake.commandRun().withTimeout(5),
+        new WaitCommand(3),
+        intake.commandRun().withTimeout(5)
+    );
+
+  .. code-tab:: c++
+
+    // TODO
+
+Adding a parameter to the ``commandRun`` method to provide the exact percentage to run the intake is easy and allows for even more flexibility.
```

Comment:
Beware of too much flexibility. In most cases it's better to accept an enum or multiple factories for a number of preset options instead of giving absolute control to whoever is calling the command.

Reply:
In real-world code I'd be inclined to agree, but this is just as an example of how command factories can be parameterized. Defining an enum and explaining the logic behind it would distract from the main lesson IMO.